---
title: "Look for contamination in metagenomics MetaPhlAn data"
author: "Roxana Hickey"
date: "Last updated December 9, 2015"
output: 
  html_document: 
    keep_md: yes
---

Adapted from analysis by James Meadow for human microbial cloud study (original script [here](https://github.com/jfmeadow/Meadow_HumanMicrobialCloud_Analysis/blob/master/manuscript_code/contamination/pb_contamination.md)).

Load data.
```{r}
library(vegan)
library(labdsv)
library(RColorBrewer)
library(gplots)

source('~/Documents/projects/dust_2015/code/meadow_pb_functions.R')
```

Read in the OTU table (relative abundance)
```{r}
## read in the genus-level OTU table
ger.meta <- read.table('~/Documents/projects/dust_2015/data/tax_shotgun_metaphlan/AAGAGGCA-AGAGGATA_S4cat_R1_001.fastq.metaphlan2_merged_meta_L6.txt', header = T, sep = '\t', row.names = 1)
ger.meta <- t(ger.meta)

colnames(ger.meta) <- gsub('k__', '', colnames(ger.meta))
colnames(ger.meta) <- gsub('p__', '', colnames(ger.meta))
colnames(ger.meta) <- gsub('c__', '', colnames(ger.meta))
colnames(ger.meta) <- gsub('o__', '', colnames(ger.meta))
colnames(ger.meta) <- gsub('f__', '', colnames(ger.meta))
colnames(ger.meta) <- gsub('g__', '', colnames(ger.meta))
colnames(ger.meta) <- gsub('Other', '', colnames(ger.meta))

ger.meta.tax <- colnames(ger.meta)

## read in master sample ID key
sid.key <- read.table('~/Documents/projects/dust_2015/data/sampleid_master_key.txt',
                      sep = '\t', header = T, row.names = NULL)
rownames(ger.meta) %in% sid.key$metaphlan_barcode

## replace sample IDs to match 16S data
ger.meta.orig <- ger.meta
rownames(ger.meta) <- sid.key$uparse_sampleid[match(rownames(ger.meta.orig), sid.key$metaphlan_barcode)]

## read in mapping file with metadata
ger.map.orig <- read.table('~/Documents/projects/dust_2015/data/map_files/151125_map_edit_finalchems_crack_plus.txt',
                           header = TRUE, sep = '\t', row.names = 1, comment.char = '')
ger.map.orig$SampleID <- row.names(ger.map.orig)

## define samples to ignore due to duplication
sample.ignore <- c('oneuL120a1', 'oneuL120a2', 'oneuL125a1', 'oneuL1402', 'oneuLh2412', 'oneuLs2101')

## make reduced OTU table
ger.meta <- ger.meta[!(row.names(ger.meta) %in% sample.ignore),]

## reduce mapping file to samples in MetaPhlAn dataset
ger.meta.map <- ger.map.orig[ger.map.orig$SampleID %in% rownames(ger.meta),]

## check that all sample names match
all(row.names(ger.meta.map) %in% row.names(ger.meta))

dim(ger.meta.map)
dim(ger.meta)

## reorder map to match OTU table
ger.meta.map <- ger.meta.map[row.names(ger.meta), ]
identical(row.names(ger.meta), row.names(ger.meta.map))

## see how many taxa there are
length(ger.meta.tax)
```

Plant and mitochondrial sequences are not present in the MetaPhlAn data, so we do not need to remove them manually as we did with the 16S rRNA data. Next, we'll make a taxonomy data frame for indexing.
```{r}
taxo <- makeTaxo(ger.meta.tax, ger.meta)
head(taxo)
```

Get rid of empty OTUs to reduce computing demand.
```{r}
taxo <- taxo[colnames(ger.meta), ]
head(taxo)

dim(taxo)
dim(ger.meta)

identical(row.names(taxo), colnames(ger.meta))
```

Quick reference vector of names for figures and whatnot. Some genus, families, etc are empty, so search for the finest taxonomic level and use that name.
```{r}
consensus <- apply(taxo[, 1:6], 1, cons)
consensus[1:10]
```

Some indexing and metadata stuff for figures and such.
```{r}
ger.meta.map$pch <- 16
table(ger.meta.map$SpaceTypeBioBE)

control <- which(ger.meta.map$SpaceTypeBioBE == 'neg.control')
samples <- which(ger.meta.map$SpaceTypeBioBE != 'neg.control')

spaces <- unique(ger.meta.map$SpaceTypeBioBE)
spaces <- spaces[-10]

spaces.counts <- list()
for(i in 1:length(spaces)){
  spaces.counts[[i]] <- which(ger.meta.map$SpaceTypeBioBE == spaces[i])
}

ger.meta.map$pch2 <- ger.meta.map$pch
ger.meta.map$pch2[ger.meta.map$SpaceTypeBioBE == 'neg.control'] <- 17

## custom color palette
mycol.10 <- c("#0000A9", "#0049FF", "#00A4DE", "turquoise", "chartreuse3", 
              "#FFD701", "#FF9500", "#FF3300", "darkmagenta", "slategrey")
names(mycol.10) <- c('building.support', 'circulation', 'classroom', 'gym', 'laundry', 
                     'lockers', 'office', 'pool', 'restroom', 'neg.control')

ger.meta.map$bg <- 'gray60'

for(i in 1:length(spaces)){
  ger.meta.map$bg[spaces.counts[[i]]] <- mycol.10[i]  
}
```

Make dissimilarity matrix with both Canberra and Bray-Curtis, and then make NMDS and PCoA ordinations. The Bray-Curtis will be more useful for identifying the influence of the most abundant taxa, while Canberra has most often been used for analysis in our previous studies. Here, Canberra has high stress, so is not used much. Also quick plot to make sure that BC captures differences between groups.
```{r fig.width = 12, fig.height = 6}
can.meta <- vegdist(ger.meta, 'canberra')
bc.meta <- vegdist(ger.meta)

## NMDS
nmds.can.meta <- nmds(can.meta)
nmds.bc.meta <- nmds(bc.meta)

par(mfrow = c(1,2))
plot(nmds.can.meta$points, col = ger.meta.map$SpaceTypeBioBE, type = "n",
     main = 'NMDS (Canberra)', xlab = 'NMDS 1', ylab = 'NMDS 2')
points(nmds.can.meta$points[control, ], pch = 17, col = ger.meta.map$bg[control])
for(i in 1:length(spaces)){
  points(nmds.can.meta$points[spaces.counts[[i]], ], 
         pch = 16, col = ger.meta.map$bg[spaces.counts[[i]]])
}

plot(nmds.bc.meta$points, col = ger.meta.map$SpaceTypeBioBE, type = "n",
     main = 'NMDS (Bray-Curtis)', xlab = 'NMDS 1', ylab = 'NMDS 2')
points(nmds.bc.meta$points[control, ], pch = 17, col = ger.meta.map$bg[control])
for(i in 1:length(spaces)){
  points(nmds.bc.meta$points[spaces.counts[[i]], ], 
         pch = 16, col = ger.meta.map$bg[spaces.counts[[i]]])
}

## PCoA
pcoa.can.meta <- cmdscale(can.meta, eig = TRUE)
pcoa.bc.meta <- cmdscale(bc.meta, eig = TRUE)

par(mfrow = c(1,2))
plot(pcoa.can.meta$points, col = ger.meta.map$SpaceTypeBioBE, type = "n",
     main = 'PCoA (Canberra)', xlab = 'PCoA 1', ylab = 'PCoA 2')
points(pcoa.can.meta$points[control, ], pch = 17, col = ger.meta.map$bg[control])
for(i in 1:length(spaces)){
  points(pcoa.can.meta$points[spaces.counts[[i]], ], 
         pch = 16, col = ger.meta.map$bg[spaces.counts[[i]]])
}

plot(pcoa.bc.meta$points, col = ger.meta.map$SpaceTypeBioBE, type = "n",
     main = 'PCoA (Bray-Curtis)', xlab = 'PCoA 1', ylab = 'PCoA 2')
points(pcoa.bc.meta$points[control, ], pch = 17, col = ger.meta.map$bg[control])
for(i in 1:length(spaces)){
  points(pcoa.bc.meta$points[spaces.counts[[i]], ], 
         pch = 16, col = ger.meta.map$bg[spaces.counts[[i]]])
}
```

Find the 10 most abundant taxa. Use 10 for plotting, and take out those contaminants. Notably in this case, the only bacterial taxa found in the negative control are Propionibacterium (genus) and Propionibacteriaceae (family)
```{r}
## top 10 taxa
control.taxa.10 <- rev(sort(ger.meta[control, ]))[1:10]/sum(ger.meta[control, ])
sample.taxa.10 <- rev(sort(colSums(ger.meta[-control, ])))[1:10]/sum(ger.meta[-control, ])

control.taxa.10['other'] <- 1 - sum(control.taxa.10)
sample.taxa.10['other'] <- 1 - sum(sample.taxa.10)

top.10 <- cbind(rev(control.taxa.10), rev(sample.taxa.10))
# mids <- barplot(top.10, col = c('gray30', rep('gray70', 10)), border = 'gray30', space = 1)

con.cum.10 <- cumsum(rev(control.taxa.10))
sam.cum.10 <- cumsum(rev(sample.taxa.10))
```

Only Propionibacterium/Propionibacteriaceae are present in the negative control. In the samples, it accounts for just under 12% of the data.
```{r fig.width = 12, fig.height = 6}
rev(sort(colSums(ger.meta)))[1:10]

c1 <- ger.meta[, 'Bacteria;Actinobacteria;Actinobacteria;Actinomycetales;Propionibacteriaceae;']
c2 <- ger.meta[, 'Bacteria;Actinobacteria;Actinobacteria;Actinomycetales;Propionibacteriaceae;Propionibacterium']

par(mfrow = c(1,2))
plot(c1 ~ nmds.bc.meta$points[, 1], type = "n", xlab = 'NMDS 1', ylab = 'Propionibacteriaceae (family best resolution)')
points(c1[control] ~ nmds.bc.meta$points[control, 1], pch = 17, bg = ger.meta.map$bg[control])
for(i in 1:length(spaces)){
  points(c1[spaces.counts[[i]]] ~ nmds.bc.meta$points[spaces.counts[[i]], 1], 
         pch = 16, col = ger.meta.map$bg[spaces.counts[[i]]])
}

plot(c2 ~ nmds.bc.meta$points[, 1], type = "n", xlab = 'NMDS 1', ylab = 'Propionibacterium (genus best resolution)')
points(c2[control] ~ nmds.bc.meta$points[control, 1], pch = 17, bg = ger.meta.map$bg[control])
for(i in 1:length(spaces)){
  points(c2[spaces.counts[[i]]] ~ nmds.bc.meta$points[spaces.counts[[i]], 1], 
         pch = 16, col = ger.meta.map$bg[spaces.counts[[i]]])
}
```

Now take out those most abundant in pcr controls.
```{r}
dim(ger.meta)

## reduce taxonomy table
ger.meta.nc <- ger.meta[, -which(colnames(ger.meta) %in% names(control.taxa.10[1:2]))]
ger.meta.nc <- ger.meta.nc[rowSums(ger.meta.nc) > 0, ]

sort(rowSums(ger.meta.nc))

taxo.nc <- makeTaxo(ger.meta.tax[ger.meta.tax %in% colnames(ger.meta.nc)], ger.meta.nc)
consensus.nc <- apply(taxo.nc[, 1:6], 1, cons)

## reduce map table
ger.meta.map.nc <- ger.meta.map[rownames(ger.meta.map) %in% rownames(ger.meta.nc),]
```

Make same NMDS objects to compare.
```{r fig.width = 12, fig.height = 6}
bc.meta.nc <- vegdist(ger.meta.nc)
bc.meta <- vegdist(ger.meta)

## NMDS
nmds.bc.meta.nc <- nmds(bc.meta.nc)
nmds.bc.meta <- nmds(bc.meta)

par(mfrow = c(1,2))
plot(nmds.bc.meta$points, type = 'n', main = 'Full dataset', xlab = 'NMDS 1', ylab = 'NMDS 2')
points(nmds.bc.meta$points[control, ], pch = 17, col = ger.meta.map$bg[control])
points(nmds.bc.meta$points[samples, ], pch = 16, col = ger.meta.map$bg[samples])

plot(nmds.bc.meta.nc$points, type = 'n', main = 'Propis removed', xlab = 'NMDS 1', ylab = 'NMDS 2')
points(nmds.bc.meta.nc$points, pch = 16, col = ger.meta.map.nc$bg)

## PCoA
pcoa.bc.meta.nc <- cmdscale(bc.meta.nc, eig = TRUE)
pcoa.bc.meta <- cmdscale(bc.meta, eig = TRUE)

par(mfrow = c(1,2))
plot(pcoa.bc.meta$points, type = 'n', main = 'Full dataset', xlab = 'PCoA 1', ylab = 'PCoA 2')
points(pcoa.bc.meta$points[control, ], pch = 17, col = ger.meta.map$bg[control])
points(pcoa.bc.meta$points[samples, ], pch = 16, col = ger.meta.map$bg[samples])

plot(pcoa.bc.meta.nc$points, type = 'n', main = 'Propis removed', xlab = 'PCoA 1', ylab = 'PCoA 2')
points(pcoa.bc.meta.nc$points, pch = 16, col = ger.meta.map.nc$bg)
```

Run some stats to see if space type explains dissimilarity
```{r}
adonis(bc.meta.nc ~ ger.meta.map.nc$SpaceTypeBioBE)
adonis(bc.meta ~ ger.meta.map$SpaceTypeBioBE)
```

Make heatmaps with and without Propi removed
```{r fig.width = 8, fig.height = 8}
## with Propi
## subset to top 25 taxa
pick <- order(colSums(ger.meta), decreasing=TRUE)[1:25]

identical(row.names(ger.meta), row.names(ger.meta.map))
mycol <- mycol.10[match(ger.meta.map$SpaceTypeBioBE, names(mycol.10))]

# png('~/Documents/projects/dust_2015/figures/ger_meta_heatmap_top25_genus_withPropi.png', width = 7, height = 6, res = 300, units = 'in', pointsize = 8)
par(xpd = TRUE)
heatmap.2(t(ger.meta[,pick]), trace = 'none',
          col = colorRampPalette(brewer.pal(9, 'YlGnBu'))(100), margin = c(6,10),
          ColSideColors = mycol,
          density.info = 'none', 
          keysize = 1, key.title = '', key.xlab = 'Proportion',
          labCol = ger.meta.map[row.names(ger.meta), 'Description'],
          labRow = consensus[colnames(ger.meta[,pick])],
          xlab = 'Sample', ylab = 'Taxon')
legend(0.9, 1.1, bty = 'n', cex = 0.7, title = 'Space Type',
       legend = c('building.support', 'circulation', 'classroom', 'gym', 'laundry', 
                  'lockers', 'office', 'pool', 'restroom', 'neg.control'),
       fill = mycol.10, border = 'white')
# dev.off()

## no Propi
## subset to top 25 taxa
pick <- order(colSums(ger.meta.nc), decreasing=TRUE)[1:25]

identical(row.names(ger.meta.nc), row.names(ger.meta.map.nc))
mycol <- mycol.10[match(ger.meta.map.nc$SpaceTypeBioBE, names(mycol.10))]

# png('~/Documents/projects/dust_2015/figures/ger_meta_heatmap_top25_genus_noPropi.png', width = 7, height = 6, res = 300, units = 'in', pointsize = 8)
par(xpd = TRUE)
heatmap.2(t(ger.meta.nc[,pick]), trace = 'none',
          col = colorRampPalette(brewer.pal(9, 'YlGnBu'))(100), margin = c(6,10),
          ColSideColors = mycol,
          density.info = 'none', 
          keysize = 1, key.title = '', key.xlab = 'Proportion',
          labCol = ger.meta.map[row.names(ger.meta.nc), 'Description'],
          labRow = consensus.nc[colnames(ger.meta.nc[,pick])],
          xlab = 'Sample', ylab = 'Taxon')
legend(0.9, 1.1, bty = 'n', cex = 0.7, title = 'Space Type',
       legend = c('building.support', 'circulation', 'classroom', 'gym', 'laundry', 
                  'lockers', 'office', 'pool', 'restroom', 'neg.control'),
       fill = mycol.10, border = 'white')
# dev.off()
```

Loose ends: calculate PCoA 'variance explained', reduce color palette
```{r}
## Use ape::pcoa to obtain R^2-like ratios
pcoa.meta.rval <- ape::pcoa(vegdist(ger.meta.nc))

## R^2-like ratio for first two axes
pcoa.meta.rval$values$Rel_corr_eig[1:2]

## reduce mycol.10 to mycol.9 (neg control no longer needed)
mycol.9 <- mycol.10[-10]
```

Clean up, save data
```{r}
rm(top.10, can.meta, con.cum.10, control.taxa.10, i, nmds.can.meta, sam.cum.10, sample.taxa.10, spaces.counts)
save.image('~/Documents/projects/dust_2015/results/otu_setup/ger_rm_contaminants_meta.RData')
```